Fix Android emulator CI job failures - 13m wasted on bugs, Gradle times out

## Problem

Android emulator tests in GitHub Actions are failing after 15 minutes with **ZERO test output**. Timeline analysis reveals the job wastes 13+ minutes on avoidable issues before Gradle times out silently.

### Recent Failure
- **Job:** https://github.com/kaeawc/auto-mobile/actions/runs/20825010890/job/59824081972
- **PR:** #441
- **Status:** CANCELLED after 15m14s
- **Tests Run:** 0 (Gradle produced zero output for 5m44s before timeout)

---

## Complete Timeline Analysis

| Phase | Duration | % of Total | Status |
|-------|----------|------------|--------|
| Runner Setup | 1s | 0.1% | ✅ Normal |
| Git Checkout & Actions | 23s | 2.5% | ✅ Normal |
| Dependencies (Bun/Java) | 20s | 2.2% | ✅ Normal |
| **Android SDK/Emulator Install** | **6m03s** | **39.6%** | 🔴 **BROKEN** |
| First Emulator Boot (Snapshot) | 43s | 4.7% | ⚠️ Avoidable |
| Second Emulator Boot (From Snapshot) | 20s | 2.2% | ✅ Normal |
| APK Installation | 7s | 0.8% | ✅ Normal |
| **Daemon Warm-up (FAILED)** | **1m19s** | **8.6%** | 🔴 **BROKEN** |
| **Test Execution (TIMEOUT)** | **5m44s** | **37.5%** | 🔴 **BROKEN** |
| Cleanup | 13s | 1.4% | ✅ Normal |
| **TOTAL** | **15m14s** | **100%** | ❌ FAILED |

**Time wasted on bugs: 13m06s (86% of job time)**

---

## Detailed Timeline Breakdown

### ✅ Phase 1-3: Normal Setup (44 seconds)
**17:05:50 - 17:06:34**
- Runner provisioning
- Git checkout
- Setup Bun/NPM packages
- Setup Java 21.0.9+10

**Status:** Working as expected

---

### 🔴 Phase 4: Android SDK Installation - **6m03s WASTED** (Bug #1)
**17:06:34 - 17:12:37** (6m03s)

**What happened:**
```
17:06:33: Start sdkmanager --install emulator
17:06:33 - 17:12:36: Installing emulator (SILENT for 6 minutes)
17:12:36: Create AVD
17:12:37: AVD created
```

**Root Cause:** AVD cache is MISSING/NOT WORKING

**Evidence:**
- Both emulator boots ran (first boot for snapshot creation + second boot from snapshot)
- This proves `steps.avd-cache.outputs.cache-hit != 'true'` (cache miss)
- Cache should skip first boot entirely if working

**Why Cache Miss:**
- Cache key may be too specific: `avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}`
- Cache might not persist between workflow runs
- First run on this PR, so cache doesn't exist yet

**Impact:** 6m03s downloading/installing emulator from scratch

---

### ⚠️ Phase 5: First Emulator Boot - 43s (Avoidable if cache works)
**17:12:37 - 17:13:20** (43s)

**What happened:**
```
17:12:37: Start emulator (cold boot, no snapshot exists)
17:12:47: Begin polling sys.boot_completed (every 2s)
17:13:20: Boot completed, unlock screen (keyevent 82)
17:13:20: Kill emulator to save snapshot
```

**Status:** Working as designed, but should be skipped when AVD cache hits

**Impact:** 43s that could be saved with proper AVD caching

---

### ✅ Phase 6-7: Second Boot & APK Install (27 seconds)
**17:13:20 - 17:13:47** (27s)

**What happened:**
```
17:13:30: Start emulator from snapshot
17:13:32: Snapshot loaded successfully (1753ms)
17:13:40: Boot completed
17:13:42: Install accessibility-service APK (24MB)
17:13:47: APK installed successfully
```

**Status:** Working as expected - snapshot boot is fast (20s)

---

### 🔴 Phase 8: Daemon Warm-up - **1m19s FAILED** (Bug #2)
**17:13:47 - 17:15:06** (1m19s, 30 failed attempts)

**What happened:**
```
17:13:47: Start daemon warm-up
17:13:49: Daemon started (PID 3855, port 3000)
17:13:49 - 17:15:06: Polling daemon 30 times (every 2s)
  Daemon shows: "availableDevices": 1, "totalDevices": 1
  But warmup check NEVER succeeds
17:15:06: ⚠️ Gave up after 30 attempts, proceeding anyway...
```

**Root Cause:** BROKEN GREP PATTERN in `scripts/android/run-emulator-tests.sh:339`

**The Bug:**
```bash
# Line 339 - BROKEN CODE:
if echo "$RESULT" | grep -q '"availableDevices":[^0]'; then
```

**Actual JSON Output:**
```json
{
  "availableDevices": 1,
  "totalDevices": 1
}
```
NOTE: There is a SPACE after the colon in the JSON!

**Why It Fails:**
- Pattern `"availableDevices":[^0]` expects NO space after colon
- Actual JSON has space: `"availableDevices": 1`
- Grep never matches, so loop always fails

**Impact:** Wastes 1m19s (30 attempts × 2s), then proceeds anyway with warning

---

### 🔴 Phase 9: Test Execution - **5m44s TIMEOUT** (Bug #3 or #4)
**17:15:06 - 17:20:50** (5m44s)

**What happened:**
```
17:15:06: Start Gradle: ./gradlew :playground:app:test --stacktrace
17:15:06 - 17:20:50: COMPLETE SILENCE (no output for 5m44s)
17:20:50: ##[error]The operation was canceled.
```

**Root Cause:** Unknown - possibilities:

**Theory #1: Gradle Downloading Dependencies (No Gradle Cache)**
- We don't cache Gradle dependencies
- Gradle might be silently downloading for 5+ minutes
- No progress output because we only use `--stacktrace`, not `--info`

**Theory #2: Gradle Waiting for Daemon (Related to Bug #2)**
- Daemon warmup failed but proceeded anyway
- Gradle tests might be waiting for daemon connection
- Timeout after 5m44s when it never connects

**Theory #3: Gradle Daemon Startup Issue**
- Gradle daemon itself might be stuck initializing
- No output during startup phase

**Impact:** Job fails without running ANY tests

---

### ✅ Phase 10: Cleanup (13 seconds)
**17:20:50 - 17:21:03**

**What happened:**
- Upload artifacts (no files - tests didn't run)
- Publish test report (0 tests found)
- Kill orphan processes

**Status:** Working as expected

---

## Root Cause Analysis

### 🔴 **Bug #1: AVD Cache Not Working (6m03s wasted)**

**Location:** `.github/actions/android-emulator/action.yml:57-64`

**Current Code:**
```yaml
- name: AVD cache
  uses: actions/cache@v4
  id: avd-cache
  with:
    path: |
      ~/.android/avd/*
      ~/.android/adb*
    key: avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}
```

**Issues:**
- Cache is missing on every run (both emulator boots happen)
- Cache key might be too specific
- No cache restore-keys fallback
- Cache might not be saved properly

**Evidence:**
- Timeline shows BOTH boots ran:
  - First boot: 17:12:37 - 17:13:20 (snapshot creation)
  - Second boot: 17:13:30 - 17:13:40 (from snapshot)
- If cache hit, first boot would be skipped

---

### 🔴 **Bug #2: Daemon Warmup Grep Pattern Broken (1m19s wasted)**

**Location:** `scripts/android/run-emulator-tests.sh:339`

**Current Code:**
```bash
if echo "$RESULT" | grep -q '"availableDevices":[^0]'; then
  print_success "Daemon is ready and devices are available"
  DAEMON_READY=true
  break
```

**Issue:** Pattern expects no space after colon in JSON

**Actual JSON:**
```json
{
  "availableDevices": 1,
  "totalDevices": 1
}
```

**Why It Fails:**
- `"availableDevices":[^0]` only matches `"availableDevices":1`
- Real output has space: `"availableDevices": 1`
- Grep never finds a match
- Loop retries 30 times (1m19s) then gives up

**Fix:**
```bash
if echo "$RESULT" | grep -q '"availableDevices": [1-9]'; then
```

Or better yet, use `jq`:
```bash
if echo "$RESULT" | jq -e '.content[0].text | fromjson | .availableDevices > 0' &>/dev/null; then
```

---

### 🔴 **Bug #3: Missing Gradle Cache (possibly 5m44s wasted)**

**Location:** Workflow doesn't include Gradle caching

**Current Pattern:** ❌ Missing
```yaml
# NO Gradle cache at all!
```

**Recommended Pattern:** ✅ Should have
```yaml
- name: Setup Gradle
  uses: gradle/actions/setup-gradle@v4
```

**Impact:**
- Gradle downloads all dependencies on every run
- No dependency caching between runs
- Possibly explains 5m44s silent timeout (downloading dependencies)

---

### 🔴 **Bug #4: Missing `-no-snapshot-save` Flag**

**Location:** `.github/actions/android-emulator/action.yml:88-98`

**Current Code:**
```yaml
- name: Run Emulator
  uses: reactivecircus/android-emulator-runner@v2.35.0
  with:
    emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
    # Missing: -no-snapshot-save
```

**Issue:** Second emulator boot might overwrite the cached AVD snapshot

**Recommended:**
```yaml
emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
```

---

### ⚠️ **Bug #5: Insufficient Gradle Logging**

**Location:** Workflow uses only `--stacktrace`

**Current:**
```bash
./gradlew :playground:app:test --stacktrace
```

**Issue:** No progress output, can't debug hangs

**Recommended:**
```bash
./gradlew :playground:app:test --stacktrace --info
```

Or for debugging:
```bash
./gradlew :playground:app:test --stacktrace --debug
```

---

## Comparison: Current vs Recommended Pattern

### ❌ **Current Pattern:**
```yaml
steps:
  - uses: actions/checkout@v4
  # No Gradle cache ❌

  - name: AVD cache  # ✅ Good!
    uses: actions/cache@v4
    with:
      path: ~/.android/avd/*
      key: avd-${{ runner.os }}-api-${{ inputs.api_level }}

  - name: Create AVD (if cache miss)  # ✅ Good!
    if: steps.avd-cache.outputs.cache-hit != 'true'
    uses: reactivecircus/android-emulator-runner@v2

  - name: Run Tests
    uses: reactivecircus/android-emulator-runner@v2
    with:
      emulator-options: -no-window ...  # ❌ Missing -no-snapshot-save
      script: ./gradlew test --stacktrace  # ❌ No --info logging
```

### ✅ **Recommended Pattern:**
```yaml
steps:
  - uses: actions/checkout@v4

  # ✅ Add Gradle cache
  - name: Setup Gradle
    uses: gradle/actions/setup-gradle@v4

  - name: AVD cache
    uses: actions/cache@v4
    id: avd-cache
    with:
      path: |
        ~/.android/avd/*
        ~/.android/adb*
      key: avd-${{ runner.os }}-api-${{ matrix.api-level }}
      restore-keys: |
        avd-${{ runner.os }}-

  - name: Create AVD (if cache miss)
    if: steps.avd-cache.outputs.cache-hit != 'true'
    uses: reactivecircus/android-emulator-runner@v2
    with:
      api-level: ${{ matrix.api-level }}
      emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim
      script: echo "AVD snapshot generated"

  - name: Run Tests
    uses: reactivecircus/android-emulator-runner@v2
    with:
      api-level: ${{ matrix.api-level }}
      emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim
      script: ./gradlew test --stacktrace --info
```

---

## Fixes Required

### **Fix #1: Add Gradle Caching**

**File:** `.github/workflows/pull_request.yml`

**Add before emulator steps:**
```yaml
- name: Setup Gradle
  uses: gradle/actions/setup-gradle@v4
```

**Impact:** Save 5+ minutes on dependency downloads

---

### **Fix #2: Fix Daemon Warmup Grep Pattern**

**File:** `scripts/android/run-emulator-tests.sh:339`

**Change from:**
```bash
if echo "$RESULT" | grep -q '"availableDevices":[^0]'; then
```

**Change to:**
```bash
if echo "$RESULT" | grep -q '"availableDevices": [1-9]'; then
```

**Or use jq for robustness:**
```bash
if echo "$RESULT" | jq -e '.content[0].text | fromjson | .availableDevices > 0' &>/dev/null; then
```

**Impact:** Save 1m19s on every run, daemon warmup will actually work

---

### **Fix #3: Add AVD Cache Restore Keys**

**File:** `.github/actions/android-emulator/action.yml:57-64`

**Add restore-keys:**
```yaml
- name: AVD cache
  uses: actions/cache@v4
  id: avd-cache
  with:
    path: |
      ~/.android/avd/*
      ~/.android/adb*
    key: avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}
    restore-keys: |
      avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-
      avd-${{ runner.os }}-${{ inputs.api_level }}-
      avd-${{ runner.os }}-
```

**Impact:** Better cache hit rate, save 6m03s when cache hits

---

### **Fix #4: Add -no-snapshot-save Flag**

**File:** `.github/actions/android-emulator/action.yml:97`

**Change from:**
```yaml
emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
```

**Change to:**
```yaml
emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
```

**Impact:** Prevent overwriting cached AVD snapshot

---

### **Fix #5: Add Gradle Logging**

**File:** `.github/actions/android-emulator/action.yml` or test scripts

**Change from:**
```bash
./gradlew :playground:app:test --stacktrace
```

**Change to:**
```bash
./gradlew :playground:app:test --stacktrace --info
```

**Impact:** Visibility into Gradle hangs, easier debugging

---

### **Fix #6: Increase Timeout (Temporary)**

**File:** `.github/workflows/pull_request.yml:295`

**Change from:**
```yaml
timeout-minutes: 15
```

**Change to:**
```yaml
timeout-minutes: 20
```

**Impact:** Temporary mitigation while debugging Gradle timeout

---

## Expected Results After Fixes

### **Before (Current - Broken):**
```
Total: 15m14s (TIMEOUT, 0 tests run)
├─ SDK Install: 6m03s (cache miss)
├─ First Boot: 43s (snapshot creation)
├─ Second Boot: 20s
├─ Daemon Warmup: 1m19s (FAILED)
└─ Gradle Test: 5m44s (TIMEOUT)
```

### **After (All Fixes Applied):**
```
Total: ~3-4 minutes (PASSING)
├─ SDK Install: SKIPPED (cache hit)
├─ First Boot: SKIPPED (cache hit)
├─ Second Boot: 20s (from cached snapshot)
├─ Daemon Warmup: 4-10s (WORKING - finds device on 2nd-5th attempt)
└─ Gradle Test: 2-3m (actual tests run with cached dependencies)
```

**Time Saved: 11-12 minutes per run (74-79% faster)**

---

## Investigation Questions

### **Question #1: Why is AVD cache missing?**

**Investigate:**
- Check if cache is being saved at end of job
- Verify cache key matches between runs
- Check GitHub Actions cache retention policy
- Look at cache storage limits

**Debug:**
```yaml
- name: Debug cache
  run: |
    echo "Cache hit: ${{ steps.avd-cache.outputs.cache-hit }}"
    echo "Cache key: avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}"
    ls -la ~/.android/avd/ || echo "No AVD directory"
```

---

### **Question #2: Why does Gradle timeout silently?**

After fixing daemon warmup (#2) and adding Gradle cache (#1) and logging (#5):

**If still times out, investigate:**
- Gradle daemon startup issues
- Network connectivity to Maven repos
- Gradle configuration cache issues
- Test infrastructure problems

**Debug:**
```bash
./gradlew :playground:app:test --stacktrace --info --debug
```

---

## Priority Order

### **Critical (Do First):**
1. ✅ **Fix #2: Daemon warmup grep pattern** (1 line change, saves 1m19s)
2. ✅ **Fix #5: Add Gradle logging** (1 line change, enables debugging)
3. ✅ **Fix #1: Add Gradle caching** (3 lines, saves 5+ minutes)

### **High Priority:**
4. ✅ **Fix #3: Add AVD cache restore-keys** (Better cache hit rate)
5. ✅ **Fix #4: Add -no-snapshot-save** (Prevent cache corruption)

### **Medium Priority:**
6. ⚠️ **Fix #6: Increase timeout** (Temporary mitigation)
7. 🔍 **Investigation: Why AVD cache missing** (Root cause analysis)
8. 🔍 **Investigation: Gradle timeout cause** (After other fixes)

---

## Related Issues

- #438: Skip long-running CI jobs for documentation-only PRs (addresses symptom)
- #441: PR that triggered this investigation
- This issue addresses root causes of slow/failing emulator tests

---

## Links

- **Failed Job:** https://github.com/kaeawc/auto-mobile/actions/runs/20825010890/job/59824081972
- **Workflow File:** `.github/workflows/pull_request.yml`
- **Emulator Action:** `.github/actions/android-emulator/action.yml`
- **Test Script:** `scripts/android/run-emulator-tests.sh`
- **Referenced Pattern:** https://github.com/ReactiveCircus/android-emulator-runner#usage-example

--- a/.github/actions/android-emulator/action.yml
+++ b/.github/actions/android-emulator/action.yml
@@ -41,6 +41,9 @@ runs:
         distribution: "zulu"
         java-version: "21"
 
+    - name: Setup Gradle
+      uses: gradle/actions/setup-gradle@v4
+
     - name: Verify Java version
       shell: ${{ inputs.shell }}
       run: |
@@ -62,6 +65,10 @@ runs:
           ~/.android/avd/*
           ~/.android/adb*
         key: avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}
+        restore-keys: |
+          avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-
+          avd-${{ runner.os }}-${{ inputs.api_level }}-
+          avd-${{ runner.os }}-
 
     - name: Create AVD and generate snapshot for caching
       if: steps.avd-cache.outputs.cache-hit != 'true'
@@ -94,5 +101,5 @@ runs:
         script: ../scripts/android/run-emulator-tests.sh "${{ inputs.accessibility-apk-path }}" "${{ inputs.script }}"
         working-directory: ${{ inputs.working-directory }}
         force-avd-creation: false
-        emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
+        emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
         disable-animations: false

--- a/.github/workflows/pull_request.yml
+++ b/.github/workflows/pull_request.yml
@@ -271,7 +271,7 @@ jobs:
       # Run AutoMobile tests that require emulator
       - uses: ./.github/actions/android-emulator
         with:
-          script: "./gradlew :junit-runner:test"
+          script: "./gradlew :junit-runner:test --stacktrace --info"
           working-directory: './android/'
           accessibility-apk-path: accessibility-service/build/outputs/apk/debug/accessibility-service-debug.apk
 
@@ -310,7 +310,7 @@ jobs:
       # Run AutoMobile tests that require emulator
       - uses: ./.github/actions/android-emulator
         with:
-          script: "./gradlew :playground:app:test --stacktrace"
+          script: "./gradlew :playground:app:test --stacktrace --info"
           working-directory: './android/'
           accessibility-apk-path: accessibility-service/build/outputs/apk/debug/accessibility-service-debug.apk
 

--- a/scripts/android/run-emulator-tests.sh
+++ b/scripts/android/run-emulator-tests.sh
@@ -336,11 +336,11 @@ for i in $(seq 1 $MAX_RETRIES); do
   echo "$RESULT"
 
   # Check if we have at least 1 available device (device pool is initialized)
-  if echo "$RESULT" | grep -q '"availableDevices":[^0]'; then
+  if echo "$RESULT" | grep -q '"availableDevices": [1-9]'; then
     print_success "Daemon is ready and devices are available"
     DAEMON_READY=true
     break
-  elif echo "$RESULT" | grep -q '"availableDevices":0'; then
+  elif echo "$RESULT" | grep -q '"availableDevices": 0'; then
     echo "Device pool not yet initialized (0 devices), waiting ${RETRY_DELAY}s..."
     sleep $RETRY_DELAY
   else

